### DISCUSSION TITLE
[Feature] Storage Template {album}  -> Move file automatically to subfolder

### DISCUSSION THREAD

**Discussion by Erd86** [immich-app/immich#6456] [Feature Request]
### The feature

Adding existing photos to an album, does not move the files to the album-subfolder on my disk.
Therefore I have to run "STORAGE TEMPLATE MIGRATION" job manually.
I would like, that the files will move directly to the {album} subfolder without starting the job manually.

### Platform

- [ ] Server
- [X] Web
- [ ] Mobile

**Comment by canton7** [comment]
Seconded: this is quite annoying at the moment. 

It would be great if this job were either run periodically, or automatically when an asset is added to an album.

**Comment by vpankaj97** [comment]
I have the same requirement - automatically run the job when:
 1. file is loaded into an album,
 2. album is closed / items are added to an album
 
Or a simple CRON scheduler provided in settings would also be a good option.

Note. For now, I have a CRON job scheduled using the API :
```
curl -L -X PUT '<YOUR_INSTANCE_HERE>/api/jobs/storageTemplateMigration' \
-H 'Content-Type: application/json' -H 'Accept: application/json' \
-H 'x-api-key: <YOUR_KEY_HERE>' \
-d '{"command": "start", "force": true}'
```

**Comment by skatsubo** [comment]
As currently we have Nightly Tasks configurable ([docs](https://docs.immich.app/administration/jobs-workers/)), I'm wondering if it makes sense to add/expose `storageTemplateMigration` there - turned off by default to preserve existing behavior.

<img width="1805" height="1325" alt="image" src="https://github.com/user-attachments/assets/2b623dfb-fccf-467b-af81-11105217563a" />

---
Related questions on Discord
- Confused on automating storage template migration
    - https://discord.com/channels/979116623879368755/1452068810243965010
- How is the storage template engine supposed to work? 
    - https://discord.com/channels/979116623879368755/1371777130207117344 or https://www.answeroverflow.com/m/1371777130207117344

---
Great technical discussion of the topic in [#contributing](https://discord.com/channels/979116623879368755/1071165397228855327) on Discord
- https://discord.com/channels/979116623879368755/1071165397228855327/1371828723011158113

<details><summary>Copy-paste from Discord, without markup and images</summary>

NoMachine: I wonder, when is STORAGE_TEMPLATE_MIGRATION job supposed to run automatically? I can't find it. I could only find these jobs that runs on midnight, but even when these are done, STM is not scheduled to run after any of them (unless I missed something)

  async handleNightlyJobs() {
    await this.jobRepository.queueAll([
      { name: JobName.ASSET_DELETION_CHECK },
      { name: JobName.USER_DELETE_CHECK },
      { name: JobName.PERSON_CLEANUP },
      { name: JobName.MEMORIES_CLEANUP },
      { name: JobName.MEMORIES_CREATE },
      { name: JobName.QUEUE_GENERATE_THUMBNAILS, data: { force: false } },
      { name: JobName.CLEAN_OLD_AUDIT_LOGS },
      { name: JobName.USER_SYNC_USAGE },
      { name: JobName.QUEUE_FACIAL_RECOGNITION, data: { force: false, nightly: true } },
      { name: JobName.CLEAN_OLD_SESSION_TOKENS },
    ]);
  } 

I'm looking into this, because I happen to run the Job manually, and it did a lot of changes to the fs (moving files into album folders), like if it hasn't run before automatically, so I'm not sure if this is intended.

schuhbacca: For inidivdual asset uploads, it runs after uploaded and metadata is extracted I think? There's no scheduled job I imagine, otherwise there would be no need to manually run the job after you change the template. 

schuhbacca: (This was just a quick peek at the code, could be wrong)

jrasm91: we only ever attempt to automatically run it once. that's after metadata extraction if the thing that triggered the job was an upload event.

Zeus: Feels like it should run nightly to me , idk

Zeus: I get theres risks though

NoMachine: so if you add/remove assets into albums, it doesn't get reflected in the fs

jrasm91: It does spam the FS with a bunch of requests

schuhbacca: Unless manually run, no

Zeus: Uh, why would it have any no-ops..? Wouldnt it just be rename/move calls?

jrasm91: right, but you still have to validate every path is where it is supposed to be

NoMachine: that's the think, doesn't look like is fulfilling the purpose of the storage template, since any change made to albums are not updated unless manual run, and in that case it could just had a accumulated a lot of changes, wich will result in a lot of fs calls

Zeus: We do? Why..? Isn’t that a repair page job?

Zeus: Like, what is the purpose of that? What would we do if it wasn’t where it was supposed to be?

Zeus: Maybe I’m overthinking it

jrasm91: how do you think the storage template is supposed to work?

Zeus: Well, it knows the path on disk from the database

Zeus: So it can generate the expected path for each asset and move if it doesn’t match

Zeus: I know the +1 adds a bit of complexity but I struggle to see why it would have to crawl the entire FS

jrasm91: it doesn't know if it is supposed to be at the path in the database though

Zeus: but doesn’t it calculate the new expected path based on the EXIF data and compare it to the database path?

Zeus: Surely that’s more efficient than just doing a disk lookup for each file

jrasm91: specifically when it's a duplicate and needs to have the +1 stuff it ends up needing to validate those

NoMachine: even if it does crawl the whole FS, isn't it better to do it with small batches (like once/twice daily), rather than waiting for a manual run and suddenly having to process thousands?

Zeus: Given 3 assets:
localDateTime     originalPath
2021-01-01       2021-01-01.jpg
2021-05-30     2021-05-30.jpg
2021-12-25       2021-10-01.jpg <--- needs to be moved

Zeus: I still think if a file is in the expected position to start with, it should be possible to quickly skip over it without a FS call

Zeus: I know the actual move gets more complex because you need to confirm nothing exists there before you put it in place

Zeus: but if I run the storage template job with no changes made, it really shouldn't require any FS calls as far as I can figure

jrasm91: That might be true actually. It looks like this is the only fs call it makes before actually needing to move a file, and this is only for assets that would actually need to be moved:
Image

Zeus: that makes sense. I don't know why it is so slow then lol

jrasm91: if you had a lot of duplicates it used to do this all the time, but it looks like we added something to abort early for already-moved-duplicates
Image

Zeus: Takes mine 5.5 minutes for 100k assets that seems really slow if it's majority in-memory and database transactions

jrasm91: yeah, exactly.

jrasm91: It also does do database queries one at a time

Zeus: ah, that would surely contribute to the overhead

jrasm91: for albums, it has to do three queries, one at a time for each asset
Image

NoMachine: I feel like I just stirred a hornet's nest 😂

jrasm91: and then the early abort if the source = destination is inside the moveFile call, but that's wrapped in a database lock so the lock also has to get acquired to early abort and then released and re-acquired for the next asset

jrasm91: yeah and that's why everyone stays 10ft away

Zeus: I feel like you could rewrite 90% of this in pure SQL and just select the assets that differ from the target path lol

Zeus: or we could use assets.updatedAt somehow

Zeus: though I don't think it currently covers all possible sources of a needed move

Zeus: or we just accept that it is slow 😛

NoMachine: speaking of storage template... ⁠help-desk-support⁠How is the storage template eng… 😛 
</details>

**Reply by MichiFr** [reply]
I'm currently testing storage template, here, too, like  `{{y}}/{{#if album}}{{album}}{{else}}{{y}}-{{MM}}-{{dd}}{{/if}}/{{y}}-{{MM}}-{{dd}} {{HH}}-{{mm}}-{{ss}}`.

I'm facing the same annoyances like the OP: When renaming an album, or adding or removing a photo from an album, this is not synchronized to the filesystem.

Above mentioned API call does the task of course, however as @skatsubo mentioned, an UI option would be much more elegant and avoid fiddling around with custom cron jobs.

Additonally it seems that this API call is linked to a single user and not to each and any Immich user. Which means I've to create a key for each single user and deploy several cron jobs, right?

Is there any discussion or progress regarding this topic?
